cmd/listen.go

```diff
@@ -0,0 +1,237 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/frain-dev/convoy"
+	"github.com/frain-dev/convoy/internal/pkg/socket"
+	convoyNet "github.com/frain-dev/convoy/net"
+	"github.com/frain-dev/convoy/util"
+	"github.com/gorilla/websocket"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v3"
+)
+
+var done chan interface{}
+var interrupt chan os.Signal
+
+const (
+	// Time allowed to write a message to the peer.
+	writeWait = 10 * time.Second
+
+	// Time allowed to read the next pong message from the peer.
+	pongWait = 10 * time.Second
+
+	// Send pings to peer with this period. Must be less than pongWait.
+	pingPeriod = (pongWait * 9) / 10
+)
+
+func addListenCommand(a *app) *cobra.Command {
+	var source string
+	var events string
+	var forwardTo string
+
+	cmd := &cobra.Command{
+		Use:   "listen",
+		Short: "Starts a websocket client that listens to events streamed by the server",
+		Run: func(cmd *cobra.Command, args []string) {
+			done = make(chan interface{})    // Channel to indicate that the receiverHandler is done
+			interrupt = make(chan os.Signal) // Channel to listen for interrupt signal to terminate gracefully
+
+			signal.Notify(interrupt, os.Interrupt) // Notify the interrupt channel for SIGINT
+
+			c, err := loadConfig()
+			if err != nil {
+				log.Fatal("Error loading config file:", err)
+			}
+
+			if util.IsStringEmpty(forwardTo) {
+				log.Fatal("flag forward-to cannot be empty")
+			}
+
+			listenRequest := socket.ListenRequest{
+				HostName:   c.Host,
+				DeviceID:   c.ActiveDeviceID,
+				SourceID:   source,
+				EventTypes: strings.Split(events, ","),
+			}
+
+			body, _ := json.Marshal(listenRequest)
```

Comment:
we are ignoring error from marshaling here, but go on to use / check it in the next line